build(deps): update to ACRA 5.9.x series

mikehardy opened the pull request:
## Pull Request template

## Purpose / Description

Update to current ACRA versions

@jimrod942 got us most of the way there with #9969

This finishes it with some Kotlin-migration changes related to upstream switching to Kotlin and re-organizing how they did config

## Fixes
Fixes #7805

## Approach

- Use kotlin-style variable references for 6 of the compile errors (trivial, these ones)
- Proposed patch upstream to open the method we used to override in Java - marking this draft until released in ACRA 5.9.x
- Proposed patch upstream to make default exception handler nullable
- Lots of configuration style changes as upstream config style changed again

## How Has This Been Tested?

Follow the ACRA testing guidelines in our wiki -  https://github.com/ankidroid/Anki-Android/wiki/Crash-Reports#steps-to-test - I just exercised those (and cleaned them up) recently so they should work well

## Learning (optional, can help others)
- Kotlin nullable Java interop is finicky https://github.com/ACRA/acra/pull/1053
- Kotlin allows configuration things that seem easy in Kotlin but suck when done in Java :sweat_smile:  https://github.com/ACRA/acra/issues/1048
- Kotlin specifies extensibility more precisely then Java if I understand this correctly https://github.com/ACRA/acra/pull/1033
  - So you can have "protected" and "final", or "protected" and "open", we need the new "protected open" combo here

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

mikehardy commented:
Upstream merged, will be released in ACRA 5.9, so this will go to sleep for a little bit, but has a bright future...

mikehardy commented:
Something else appears to have changed between 5.8.x and 5.9.x because when I posted my PR, I had this working with the PR's change to ACRA applied locally, and now it doesn't work on the actual release version. Will inspect later.

mikehardy commented:
Tracking upstream https://github.com/ACRA/acra/issues/1048

mikehardy commented:
Still working through this one from time to time.
Found another item from new releases - apparently robolectric does not set a default uncaught exception handler, and that's acceptable as Thread.defaultUncaughtExceptionHandler is nullable per android docs

But ACRA was expecting one, upstream PR here, hopefully in a future release, maybe 5.9.2 https://github.com/ACRA/acra/pull/1053

mikehardy commented:
Okay, this is ready for review but it won't pass testing until 5.9.2 from ACRA with a fix for null uncaughtExceptionHandler
The code here won't change a bit though

mikehardy commented:
Once this is in, a separate task should be to refactor this and pull ACRA config out to a separate object
That wasn't possible before because of the annotation style of config, but now that's no longer a factor. Should be relatively easy
New config object should be in Kotlin, then configuration gets a bit easier I think

david-allison commented on build.gradle:
```diff
@@ -6,7 +6,8 @@ buildscript {
     ext.kotlin_version = '1.6.20'
     ext.lint_version = '30.1.3'
-    ext.acra_version = '5.7.0'
+    ext.lint_version = '30.1.2'
```
This seems incorrect

mikehardy replied:
Probably screwed that up while trying to keep commits ordered and mish-mashed them during a rebase, good catch

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java:
```diff
@@ -529,6 +522,41 @@ private static Configuration getLanguageConfig(@NonNull Configuration remoteConf
+     * @param dialogEnabled boolean true to enable the dialog plugin
+     * @param toastText string to show user if ACRA begins executing
+     */
+    private org.acra.config.Configuration[] getACRAPluginConfigurations(boolean dialogEnabled, String toastText) {
```
`@KotlinCleanup` - ensure callers use named argument for boolean

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/HelpDialog.kt:
```diff
@@ -259,13 +259,13 @@ object HelpDialog {
+                val builder = AnkiDroidApp.getInstance().acraCoreConfigBuilder
+                val dialogConfig = DialogConfigurationBuilder()
+                dialogConfig.enabled = true
+                builder.withPluginConfigurations(DialogConfigurationBuilder().build())
```
⚠️ This seems like we're creating a builder, setting a value on the builder, then discarding it and using a new builder.

I would expect `builder.withPluginConfigurations(dialogConfig.build())`

david-allison commented on AnkiDroid/src/test/java/com/ichi2/utils/FileUtilTest.kt:
```diff
@@ -15,7 +15,7 @@
 package com.ichi2.utils
 
-import org.acra.util.IOUtils.*
+import org.acra.util.IOUtils.writeStringToFile
```
Was the build broken before this commit? I presume it needs rebasing earlier?

david-allison commented:
And unit test issues

mikehardy commented:
Now that upstream 5.9.2 is out, let me rework this one, and try to chunk the commits up much better
It does pass testing when you have the result of my PR for ACRA in there - without it, it will fail but I wasn't going to commit my temporary patch file in their package directory that I was using to develop the PR for them

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt:
```diff
@@ -238,16 +244,18 @@ object CrashReportService {
-        ACRA.getErrorReporter().putCustomData("origin", origin)
-        ACRA.getErrorReporter().putCustomData("additionalInfo", additionalInfo)
-        ACRA.getErrorReporter().handleException(e)
+        if (FEEDBACK_REPORT_NEVER != reportMode) {
```
This seemingly small change - not sending to ACRA if we are not set to report anyway - is vital for ACRA >5.9 + Robolectric. Robolectric does not define hardware constants or execute logcat, and ACRA 5.9+ is in kotlin and much less tolerant of nulls and fails like that. Our robolectric tests are in debug / FEEDBACK_REPORT_NEVER mode though, so it works.

Everything else is NF

mikehardy commented:
The commits should be a lot easier to follow now, they take a little refactoring journey from A (pre-existing Java style config) to B (new more Kotlin-y style config). 

The basic point is that you can't easily reach in and modify live configs anymore, you need to re-config the whole system now to change things. So I localized the state and re-gen the config each time. Seems to work

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt:
```diff
@@ -93,30 +93,36 @@ object CrashReportService {
+                    .withEnabled(true)
+                    .build(),
+                ToastConfigurationBuilder()
+                    .withText(toastText)
```
future cleanup (for a follow-up): can toastText be moved to ToastType enum (containing the R.string), and actually get the string here to ensure we handle translations in the same way?

mikehardy replied:
peel to issue post-merge

david-allison commented:
Really minor change requested for performance on the startup path.

Everything else can be ignored or moved to a new issue - implementer's choice. Happy to see this go as soon as the minor change is made.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt:
```diff
@@ -154,17 +164,15 @@ object CrashReportService {
-                dialogBuilder.setEnabled(false)
-                toastBuilder.setResText(R.string.feedback_auto_toast_text)
+                dialogEnabled = false
+                toastText = mApplication.getString(R.string.feedback_auto_toast_text)
```
Of course not, why would that be important? (I'm joking, good catch and I'll figure that out somehow - will need some thinking as ACRA removed the passing of context / application so they don't take resource anymore, just the dereferenced string...)

mikehardy replied:
I was unable to quickly fix this, the Application context is never updated with the system language, the only way Activities update is that they recreate themselves. I left a big comment instead that we should peel to another issue, with a maybe-useful solution sketch of having the base contexts be wrappers that allow resources/configuration updates post-hoc on language change events vs having them be frozen post-attachBaseContext

I'll have to peel it to another issue in order to get this update in so the PR doesn't decay any more

mikehardy commented:
![Ill-be-back](https://user-images.githubusercontent.com/782704/176050246-d2f90b95-ea17-4c17-88e3-d3ddc20078ec.jpg)

mikehardy commented:
Okay, it's rebased now, but:
1- the dialog language is always system language, it does not take the app language pref
2- the toast language takes that app language *but does not correctly change it*
3- the setting to never report is not working, if you trigger a test crash on a debug build (which is "never report" by default) it will trigger a toast about collecting info then show the dialog - it should not

So, the rebase wasn't too bad but the functionality isn't all there yet

mikehardy commented:
I think I've addressed all the stuff, with some useful-for-future people guidance for future folks on the peel-out of a new issue for dynamic language support

mikehardy commented:
Finally ready! Except dynamic language switching during same app run. Does take dynamic language after app restarts.

mikehardy commented:
Just added a final commit final commit that should correctly show the app version number in Acrarium, instead of the library version number. That's going to be a huge help during the beta cycle.

Also changed our init style just a bit to work around a bug I found in ACRA in a way that works with current ACRA release and the pending/unreleased bugfix

I'll leave this out for a little while hoping for review but I want this in quickly so will merge by default in a few days if not. Always happy to take issues or work on PRs in the future though in case review is post-merge, it's still useful...

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/CrashReportService.kt:
```diff
@@ -340,28 +341,15 @@ object CrashReportService {
-                val timestampMethod = report.javaClass.getDeclaredMethod("getTimestamp")
-                timestampMethod.isAccessible = true
-                val timestamp = timestampMethod.invoke(report) as Calendar
-                // Limiter ensures there is only one report for the class, so if we found it, return it
-                return timestamp.timeInMillis
-            }
-        } catch (e: Exception) {
-            Timber.w(e, "Unexpected exception checking for recent reports")
-        }
-        return -1
+            )
+            .max().orElse(-1)
```
nitpick: you get nicer kotlin extension methods if you remove `stream()`

```kotlin
        return LimiterData.load(activity).reportMetadata
            .filter { it.exceptionClass == UserSubmittedException::class.java.name }
            .maxOfOrNull { it.timestamp?.timeInMillis ?: -1L } ?: -1L
```

mikehardy replied:
Ah cool - this was actually something F43nd1r (ACRA maintainer) cooked up for me but it was java at the time and AndroidStudio "auto-Kotlin-ed" it on paste and I just naively left it. I'll ingest your suggestion after making sure I understand it thoroughly (read as: kotlin learning exercise for me) then re-push. But no other changes so if it goes green I'll merge

And hopefully we get correct app versions in acrarium after!

david-allison replied:
`{ it.exceptionClass }` is shorthand for `.filter({ it -> it.exceptionClass})`:

* If a lambda is the final parameter, it can be moved outside the brackets
* `it` is shorthand for the parameter in a unary lambda.

But, there's no requirement to make this change, only as a learning experience

mikehardy replied:
I like it - slowly filling my head with this library of `maxOfOrNull`s and such. Too many to even contemplate memorizing at the moment (https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/) ! I do love the effortless lambda + collection handling though, it is quite nice. Pushed

mikehardy commented:
I've peeled the two possible cleanups off for follow-on issues and pushed a cleaner version of our usage of the limiter data to throttle exceptions, those were the only items from first review.

Ready for merge pending second review
